synapse/storage/account_data.py

```diff
@@ -274,40 +281,46 @@ def add_account_data_txn(txn, next_id):
                 values={
                     "stream_id": next_id,
                     "content": content_json,
-                }
+                },
+                lock=False,
             )
-            txn.call_after(
-                self._account_data_stream_cache.entity_has_changed,
+
+            # it's theoretically possible for the above to succeed and the
+            # below to fail - in which case we might reuse a stream id on
+            # restart, and the above update might not get propagated. That
+            # doesn't sound any worse than the whole update getting lost,
+            # which is what would happen if we combined the two into one
+            # transaction.
+            yield self._update_max_stream_id(next_id)
+
+            self._account_data_stream_cache.entity_has_changed(
                 user_id, next_id,
             )
-            txn.call_after(self.get_account_data_for_user.invalidate, (user_id,))
-            txn.call_after(
-                self.get_global_account_data_by_type_for_user.invalidate,
+            self.get_account_data_for_user.invalidate((user_id,))
+            self.get_global_account_data_by_type_for_user.invalidate(
                 (account_data_type, user_id,)
             )
```

Comment:
I think we might need to think if its fine that the invalidation happens after another yield or not, I'm suspecting not.

Reply:
I wondered that. But given there was no guarantee how quickly after the db update the invalidation was happening before, I couldn't see how it could make any difference. The extra yield is equivalent a sleep() in the db update thread before the call_after code returns.

Unless, of course, there's a race already, in which case yes this will make it worse...